### DISCUSSION TITLE
Proper way to initiate atomWithQuery only on event?

### DISCUSSION THREAD

**Discussion by Akkuma** [pmndrs/jotai#812] [Q&A] [answered]
I have an atomWithQuery that I'd like to only run after a button click. I have it wired up and working when using an atom to determine enabled state, but when I force it to disabled `enabled: false` the loading begins, but never completes

```ts
export const search = atomWithQuery((get) => ({
    queryKey: [get(searchTerm), get(searchLocation)],
    queryFn: async ({ queryKey: [term, location]}) => {
        // Fetch logic
        
        const json = await res.json()
       // other stuff
        return json
    },
    enabled: false
}))
```

```ts
//A separate hook
const useSearch = () => {
    const setEnabledSearch = useUpdateAtom(enabledSearch)
    const startSearch = useUpdateAtom(searchResult)

    const search = useCallback(() => {
        setEnabledSearch(true)
        startSearch({ type: 'refetch'})
    }, [startSearch])

  return { search }
}
```

```ts
// Component onClick event <Comp onClick={search} />
```

**Comment by dai-shi** [comment]
While I think running on button click should be handled in an imperative way (= callback function), there's no such things now except for `refetch`...

To avoid suspending with `enabled: false`, you need to provide `initialData`. We have been struggling with this behavior, and it's not clearly documented yet. cc: @aulneau

**Reply by Akkuma** [reply]
That doesn't seem to do the trick for me setting `initialData`. If I edit the code at  https://github.com/pmndrs/jotai/blob/main/src/query/atomWithQuery.ts#L164-L166 to be attached regardless of enabled/disabled state it works. If I understand the code correctly, that without this subscribe nothing else matters as it'll never pick up on the manual refetch performed at https://github.com/pmndrs/jotai/blob/main/src/query/atomWithQuery.ts#L175-L177.

**Reply by dai-shi** [reply]
Hm, so there's some misunderstanding of mine. sorry about it. let's see how [the other thread](https://github.com/pmndrs/jotai/discussions/812#discussioncomment-1570335) goes.

If `enabled: false`, we can't `refetch`, which is expected. (not sure if it's ideal.)

**Reply by Akkuma** [reply]
I think it could make sense to allow some way of forcing a fetch regardless of if it is "enabled".

**Reply by dai-shi** [reply]
Yeah, that's probably related to what I meant with my very first line in the comment. We'd like to basically follow how useQuery in RQ works, if possible. Do you know if we can invoke fetch function if it's disabled with useQuery? How would you do it with bare RQ? (I mean, I'm not super familiar with RQ. 😛 )

**Comment by aulneau** [comment, marked as answer]
Hi there! I made a small sandbox with what I understand your goals to be: https://codesandbox.io/s/vibrant-bhaskara-c5s7f

And it seems to be functioning, please let me know if I am missing something.

**Reply by Akkuma** [reply]
Unfortunately, the problem in my case with that solution arises from:
```js
queryKey: [get(searchTerm), get(searchLocation)],
queryFn: async ({ queryKey: [term, location]}) => 
```

Both searchTerm and searchLocation are being updated through an input and that input will cause this to trigger additional fetches as it is now enabled, when I want the searching to occur only on click.

**Reply by aulneau** [reply]
right, but why can't this work?

```ts
export const search = atomWithQuery((get) => ({
    queryKey: [get(searchTerm), get(searchLocation), get(isEnabled)],
    queryFn: async ({ queryKey: [term, location]}) => {
        // Fetch logic
        
        const json = await res.json()
       // other stuff
        return json
    },
    enabled: get(isEnabled)
}))
```

**Reply by Akkuma** [reply]
That does the same thing. term & location are updated from text inputs and I'm updating the atom `onChange`. Once isEnabled is true it'll keep fetching when term or location atoms are updated.

**Reply by aulneau** [reply]
Here is a more verbose example with a similar requirement that you've described: https://codesandbox.io/s/dank-smoke-w3rcq?file=/src/App.tsx:165-196 does this help?

**Reply by aulneau** [reply]
The important thing is to disable the query when the input changes

**Reply by Akkuma** [reply]
That definitely works. Thanks! I approached it from a singular place, since I see you're disabling on a per input basis.

I did this:
```
const term = useAtomValue(searchTerm)
const location = useAtomValue(searchLocation)
const setQueryParams = useUpdateAtom(searchQueryParams)

const search = useCallback(() => {
  setQueryParams({ term, location })
  setEnabledSearch(true)
}, [term, location])
```

```
queryKey: [get(searchQueryParams)],
queryFn: async ({ queryKey: [{term, location}]})
```

I think the "easiest" recipe is update a singular atom on callback for anything that might be dynamic, so it'll only trigger appropriately through the event callback. The downside this causes additional renders as `term`, `location`, and `search` will be changing.

**Reply by Akkuma** [reply]
Using your solution I reworked it into a hook that I could reuse:
```ts
import { useUpdateAtom } from "jotai/utils"
import { useCallback } from "react"
import { isSearchEnabled } from "../atoms/search"

const useSearchInput = ({ atom }) => {
    const setAtom = useUpdateAtom(atom)
    const setIsSearchEnabled = useUpdateAtom(isSearchEnabled)

    const onChange = useCallback((e) => {
        setIsSearchEnabled(false)
        setAtom(e.target.value)
    }, [setAtom, setIsSearchEnabled])

    return {onChange}
}

export default useSearchInput
```

That removes the need for the extra atom and searching is then just:
```ts
const search = useCallback(() => {
    setEnabledSearch(true)
}, [setEnabledSearch])
```

**Reply by aulneau** [reply]
nice! you can try this, too:
```ts
import { useAtomCallback } from 'jotai/utils';
import { useCallback } from 'react';

const useSearchInput = ({ atom }) =>
  useAtomCallback<void, React.ChangeEvent<HTMLInputElement>>(
    useCallback((get, set, e) => {
      set(isSearchEnabled, false);
      set(atom, e.currentTarget.value);
    }, [atom])
  );

```
